I2P: adds remaining work for basic I2P support

chisa0a opened the pull request:
---
name: Pull Request
about: Pull Request checklist
title: ''
labels: ''
assignees: ''

---
If your PR is a work in progress, please feel free to create it and include a [WIP] tag in the PR name. We encourage everyone to PR early and often so that other developers know what you're working on.

Before submitting your PR for final review, please ensure that it:

* Includes a proper description of what problems the PR addresses, as well as a detailed explanation as to what it changes
* Explains whether/how the change is consensus breaking or breaks existing client functionality
* Contains unit tests exercising new/changed functionality
* Fully considers the potential impact of the change on other parts of the system
* Describes how you've tested the change (e.g. against Floonet, etc)
* Updates any documentation that's affected by the PR

masterindesign commented:
Wel you called Satoshi Nakamoto an asshole. Raspberry pi tree exp.  Rust . grin like bitcoin 
nimblebe humblebe and many more ...

masterindesign commented:
Do I have to explain more tricks you picked up .
Writing in full words , be on every page. Satoshi Nakamoto is a genius , you are a programmer with  authority issues . From father to daughter we claim all the blockchaindesigns you worked on.

MarketAnarchist commented:
I am so confused right now. Who called satoshi an asshole?

masterindesign commented:
No , I just heard that somewhere - it's forgotten - wasn't you

On 16 Dec 2017 22:12, "MarketAnarchist" <notifications@github.com> wrote:

> I am so confused right now. Who called satoshi an asshole?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ignopeverell/grin/pull/1#issuecomment-352212149>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AhASnspxY_jqUCTyX8s6_KGhafaXMaO4ks5tBDIogaJpZM4Q-LVb>
> .
>

e-max committed:
introduce an async version of stratum server. (#2468)

It seems that current approach stops workring after amount of parallel
connection exceeds ~500. This PR introduces an 'async' feature which
enable asyncronius, tokio based implementation of stratum server.

* fix bug with passing current_difficulty in Handler

I passed in Handler current_difficulty as an u64, so it was copied
and all later changes were ignored. In this commit I pass referece on
RwLock with current_difficulty.

* fix crash in build_block

* improve  error handling

* print error message on unknown error

* replace original server by tokio based one

* fixes after review

* reduce sleep time in mail loop to 5ms

svechinsky committed:
Feature/slate serialization (#2534)

* - Add backwards compatability
- Add hex serialization

* rustfmt

* rustfmt

* Windows Compatibility Fixes #1 (#2535)

* initial changes for windows build and unit/integration tests

* rustfmt

* wallet+store tests

* rustfmt

* fix linux daemonize

* better encapsulate file rename

* rustfmt

* remove daemonize commands

* rustfmt

* remove server start/stop commands

* add ability to drop pmmr backend files explicitly for txhashset unzip

* rustfmt

* fix pmmr tests

* rustfmt

* Windows TUI Fix (#2555)

* switch pancurses backend to win32

* revert changes to restore test

* compatibility fix + debug messages

* rustfmt

* Add content disposition for OK responses  (#2545)

* Testing http send and fixing accordingly

* add repost method into wallet owner api (#2553)

* add repost method into wallet owner api

* rustfmt

* Add ability to compare selection strategies (#2516)

Before tx creation user can estimate fee and locked amount
with different selection strategies by providing `-e` flag for
`wallet send` command.

garyyu committed:
clean-up the txhashset zip file and temp folder (#2575)

* clean-up the txhashset zip file and temp folder
* adapt test_txhashset with the new zip_read function
* fix: the wrong folder when cleaning up old zips

e-max committed:
fix deadlock between get_status and main loop (#2556)

* fix deadlock between get_status and main loop

* isolate all shared state in one struct in order to avoid deadlocks

Commit:
Remove wallet code [1.1.0] (#2593)

* Bump version to 1.1.0

* remove wallet crate

* wallet extracted + mine_block structs

* remove wallet doc

* remove unnecessary cargo deps

* rustfmt

* remove wallet from travis matrix

* move integration tests into separate crate

* rustfmt

* move integration crate to wallet

Commit:
merge from master

Commit:
rustfmt

Commit:
insert windows build support (#2626)

Commit:
windows cargo settings (#2632)

mcdallas committed:
Event callbacks for Network and Chain Events (#2598)

* Add hooks for network and chain events. Move logging to an EventLogger
* implement webhooks
* fix failing test
* remove unnecessary 'pub'
* add some metadata to the json payload
* avoid unecessary init
* resolve conflicts

Commit:
Dynamic LMDB mapsize allocation [1.1.0] (#2605)

* dynamically resize lmdb

* rustfmt

* explicitly close db before resizing

* rustfmt

* test fix

* rustfmt

* pool tests

* chain fix

* merge

* move RwLock into Store, ensure resize gives a min threshold

* rustfmt

* move locks based on testing

* rustfmt

Commit:
comment out fuzz tests for now, remove wallet crate reference (#2635)

Commit:
disable gen_gen crate (#2636)

Commit:
Fix TxHashSet file filter for Windows. (#2641) (#2643)

* Fix TxHashSet file filter for Windows.

* rustfmt

* Updating regexp

* Adding in test case

Commit:
[1.1.0] LMDB Naming consistency fix (#2656)

* allow separate db name in store creation

* rustfmt

* fixes to db paths to ensure consistency with 1.0.x

bddap committed:
make error types serializable (#2659)

Commit:
[1.1.0] Serialise sigs as raw data, not DER encoded (#2668)

* serialise signatures as raw data

* rustfmt

Commit:
[1.1.0] Additional secp primitive serialization changes (#2669)

* serialize sig, not just option(sig)

* rustfmt

* serialise pubkeys in short form

mcdallas committed:
https support for webhooks (#2660)

* add TLS support
* configurable nthreads and timeout

hashmap committed:
[1.1.0] Bring fuzz test back (#2675)

Corpus generation was updated and simplified, plus small updates

antiochp committed:
Dandelion++ Rewrite (#2628)

* reworked the dandelion rewrite (dandelion++)

* fallback to fluff/broadcast if we cannot stem the tx for any reason

* rework stem vs fluff logic during accepting tx

* cleanup docs

* add is_stem to logging

* cleanup

* rustfmt

* cleanup monitor and logging

* rework dandelion monitor to use simple cutoff for aggregation

* transition to next epoch *after* processing tx
so we fluff final outstanding txs

* fluff all txs in stempool if any are older than 30s
aggressively aggregate when we can

* fix rebase onto 1.1.0

* default config comments for Dandelion

* fix code to reflect our tests - fallback to txpool on stempool error

* log fluff and expire errors in dandelion monitor

* cleanup

* fix off by one

* cleanup

* cleanup

* various fixes

* one less clone

* cleanup

Commit:
allow specific nonce in aggsig sign single call (#2696)

Commit:
add string_or_u64 serialization (#2698)

Commit:
fix option_sig_serde (#2699)

mcdallas committed:
Disable hooks during sync (#2695)

* disable hooks during sync
* remove trigger in headers_sync

Commit:
[1.1.0] CbData Serialization change + reward output test mode (#2710)

* cb serialization and coinbase test mode

* rustfmt

Commit:
0 the 0 that should have been a 1 (#2711)

Commit:
[1.1.0] Serialize Blockfees struct consistenly with grin-wallet (#2717)

* ser blockfees consistently with grin-wallet

* rustfmt

Commit:
[1.1.0] Merge master into 1.1.0 (#2720)

* cleanup legacy "3 dot" check (#2625)

* Allow to peers behind NAT to get up to preferred_max connections (#2543)

Allow to peers behind NAT to get up to preffered_max connections

If peer has only outbound connections it's mot likely behind NAT and we should not stop it from getting more outbound connections

* Reduce usage of unwrap in p2p crate (#2627)

Also change store crate a bit

* Simplify (and fix) output_pos cleanup during chain compaction (#2609)

* expose leaf pos iterator
use it for various things in txhashset when iterating over outputs

* fix

* cleanup

* rebuild output_pos index (and clear it out first) when compacting the chain

* fixup tests

* refactor to match on (output, proof) tuple

* add comments to compact() to explain what is going on.

* get rid of some boxing around the leaf_set iterator

* cleanup

* [docs] Add switch commitment documentation (#2526)

* remove references to no-longer existing switch commitment hash

(as switch commitments were removed in ca8447f3bd49e80578770da841e5fbbac2c23cde
and moved into the blinding factor of the Pedersen Commitment)

* some rewording (points vs curves) and fix of small formatting issues

* Add switch commitment documentation

* [docs] Documents in grin repo had translated in Korean.  (#2604)

*  Start to M/W intro translate in Korean
*  translate in Korean
*  add korean translation  on intro
* table_of_content.md translate in Korean.
*  table_of_content_KR.md finish translate in Korean, start to translate State_KR.md
*  add state_KR.md & commit some translation in State_KR.md
*  WIP stat_KR.md translation
*  add build_KR.md && stratum_KR.md
*  finish translate stratum_KR.md & table_of_content_KR.md
*  rename intro.KR.md to intro_KR.md
*  add intro_KR.md file path each language's  intro.md
*  add Korean translation file path to stratum.md & table_of_contents.md
*  fix difference with grin/master

* Fix TxHashSet file filter for Windows. (#2641)

* Fix TxHashSet file filter for Windows.

* rustfmt

* Updating regexp

* Adding in test case

* Display the current download rate rather than the average when syncing the chain (#2633)

* When syncing the chain, calculate the displayed download speed using the current rate from the most recent iteration, rather than the average download speed from the entire syncing process.

* Replace the explicitly ignored variables in the pattern with an implicit ignore

* remove root = true from editorconfig (#2655)

* Add Medium post to intro (#2654)

Spoke to @yeastplume who agreed it makes sense to add the "Grin Transactions Explained, Step-by-Step" Medium post to intro.md

Open for suggestions on a better location.

* add a new configure item for log_max_files (#2601)

* add a new configure item for log_max_files

* rustfmt

* use a constant instead of multiple 32

* rustfmt

* Fix the build warning of deprecated trim_right_matches (#2662)

* [DOC] state.md, build.md and chain directory documents translate in Korean.  (#2649)

*  add md files for translation.

*  start to translation fast-sync, code_structure. add file build_KR.md, states_KR.md

* add dandelion_KR.md && simulation_KR.md for Korean translation.

*  add md files for translation.

*  start to translation fast-sync, code_structure. add file build_KR.md, states_KR.md

* add dandelion_KR.md && simulation_KR.md for Korean translation.

* remove some useless md files for translation. this is rearrange set up translation order.

*  add dot end of sentence & translate build.md in korean

*  remove fast-sync_KR.md

*  finish build_KR.md translation

*  finish build_KR.md translation

*  finish translation state_KR.md & add phrase in state.md to move other language md file

* translate blocks_and_headers.md && chain_sync.md in Korean

*  add . in chain_sync.md , translation finished in doc/chain dir.

* fix some miss typos

* Api documentation fixes (#2646)

* Fix the API documentation for Chain Validate (v1/chain/validate).  It was documented as a POST, but it is actually a GET request, which can be seen in its handler ChainValidationHandler
* Update the API V1 route list response to include the headers and merkleproof routes.  Also clarify that for the chain/outputs route you must specify either byids or byheight to select outputs.

* refactor(ci): reorganize CI related code (#2658)

Break-down the CI related code into smaller more maintainable pieces.

* Specify grin or nanogrins in API docs where applicable (#2642)

* Set Content-Type in API client (#2680)

* Reduce number of unwraps in chain crate (#2679)

* fix: the restart of state sync doesn't work sometimes (#2687)

* let check_txhashset_needed return true on abnormal case (#2684)

*  Reduce number of unwwaps in api crate  (#2681)

* Reduce number of unwwaps in api crate

* Format use section

* Small QoL improvements for wallet developers (#2651)

* Small changes for wallet devs

* Move create_nonce into Keychain trait

* Replace match by map_err

* Add flag to Slate to skip fee check

* Fix secp dependency

* Remove check_fee flag in Slate

* Add Japanese edition of build.md (#2697)

* catch the panic to avoid peer thread quit early (#2686)

* catch the panic to avoid peer thread quit before taking the chance to ban
* move catch wrapper logic down into the util crate
* log the panic info
* keep txhashset.rs untouched
* remove a warning

* [DOC] dandelion.md, simulation.md ,fast-sync.md and pruning.md documents translate in Korean. (#2678)

* Show response code in API client error message (#2683)

It's hard to investigate what happens when an API client error is
printed out

* Add some better logging for get_outputs_by_id failure states (#2705)

* Switch commitment doc fixes (#2645)

Fix some typos and remove the use of parentheses in a
couple of places to make the reading flow a bit better.

* docs: update/add new README.md badges (#2708)

Replace existing badges with SVG counterparts and add a bunch of new ones.

* Update intro.md (#2702)

Add mention of censoring attack prevented by range proofs

* use sandbox folder for txhashset validation on state sync (#2685)

* use sandbox folder for txhashset validation on state sync

* rustfmt

* use temp directory as the sandbox instead actual db_root txhashset dir

* rustfmt

* move txhashset overwrite to the end of full validation

* fix travis-ci test

* rustfmt

* fix: hashset have 2 folders including txhashset and header

* rustfmt

* 
(1)switch to rebuild_header_mmr instead of copy the sandbox header mmr 
(2)lock txhashset when overwriting and opening and rebuild

* minor improve on sandbox_dir

* add Japanese edition of state.md (#2703)

* Attempt to fix broken TUI locale (#2713)

Can confirm that on the same machine 1.0.2 TUI looks great and is broken on
the current master. Bump of `cursive` version fixed it for me.
Fixes #2676

* clean the header folder in sandbox (#2716)

* forgot to clean the header folder in sandbox in #2685

* Reduce number of unwraps in servers crate (#2707)

It doesn't include stratum server which is sufficiently changed in 1.1
branch and adapters, which is big enough for a separate PR.

* rustfmt

* change version to beta

Commit:
tweak for windows build

Commit:
fixes for windows build

Commit:
reset windows build

antiochp committed:
Only flush leaf_set if we have a prunable backend. (#2735)

ignopeverell committed:
I2P config and peer address wrapper

Adds necessary configuration bits to enable I2P in various modes,
unused for now but will soon. Also add peer address wrappers for
I2P so those addresses can be understood and exchanged between
peers. Doesn't compile yet as serde needs to be supported in
i2p-rs first.

ignopeverell committed:
Many reference fixes now that PeerAddr isn't Copy

ignopeverell committed:
Move i2p config to p2p, set capabilities

quentinlesceller committed:
CI: Do not crash on empty terminal (#2749)

* Add warning

rentenmark committed:
decrease chunk size in outputs by ids (#2747)

mcdallas committed:
CI Test Windows: bind to localhost instead of 0.0.0.0 (#2751)

Commit:
[DOC] Translation some files in doc dir.  (#2750)

*  add md files for translation.

*  start to translation fast-sync, code_structure. add file build_KR.md, states_KR.md

* add dandelion_KR.md && simulation_KR.md for Korean translation.

*  add md files for translation.

*  start to translation fast-sync, code_structure. add file build_KR.md, states_KR.md

* add dandelion_KR.md && simulation_KR.md for Korean translation.

* remove some useless md files for translation. this is rearrange set up translation order.

*  add dot end of sentence & translate build.md in korean

*  remove fast-sync_KR.md

*  finish build_KR.md translation

*  finish build_KR.md translation

*  finish translation state_KR.md & add phrase in state.md to move other language md file

* translate blocks_and_headers.md && chain_sync.md in Korean

*  add . in chain_sync.md , translation finished in doc/chain dir.

* fix some miss typos

*  start to translate dandelion.md & simulation.md in Korean.

*  [WIP] translation

* [WIP] files add.

* [WIP] dandelion simulation

*  finish pruning translation

*  doc/dandelion translation in Korean finish

*  start to translation mmr, merkle, switch commitment in Korean

*  [WIP] merkle_KR.md

*  finish translation mmr.md & merkle.md

*  delete [WIP]switch_commitment_KR.md

*  add pow_KR.md for translation

*  finish translation grin4bitcoiners

*  fix for merge

*  fix for merge

*  fix for merge

* ,

* POW & grin4bitcoiner.md translated in Korean.

* fix some typo and cargo.lock

antiochp committed:
Pull initial chain compaction out of init() and into the syncer (#2738)

antiochp committed:
[1.1.0] Support variable size PMMR (optional size_file) (#2734)

* Introduce optional size_file.
Support fixed size data file via an optional elmt_size.
Support variable size data file via optional size_file.

* remember to release the size_file

* fix scoping for windows support

ignopeverell committed:
I2P connection init and addr/key handling

ignopeverell committed:
Starting listener impl and The Great Wrapping

quentinlesceller committed:
Cleanup db directory after tests (#2752)

* Cleanup db directory after tests

* Fix clean output dir windows

* Remove behind chain tests

Commit:
Use next available key on duplicate coinbase commitment (#2737)

If a coinbase commitment hits a duplication and there is no transactions
to mine, it is possible that coinbase cannot be built for quite a long
time. This change tries to build coinbase with an empty key identifier
immediately in this case. The listening wallet will then use the next
available key to build a coinbase commitment, which is different from
the previous one and is unlikely to hit the duplication.

Commit:
merge master into 1.1.0

Commit:
rustfmt

Commit:
version number update

ignopeverell committed:
[1.1.0 -> Master] Merge 1.1.0 with master changes back into master

Commit:
Add helpful information if user tries to run wallet command (#2759)

* Add warning message if user tries to run wallet command

* rustfmt

antiochp committed:
pass peer_info around rather than peer_addr (includes protocol version) (#2761)

Commit:
make field in tx_body public (#2762)

ignopeverell committed:
Listener impl more or less done, but untested

ignopeverell committed:
rustfm

garyyu committed:
Get back the lost code of #2575 (#2771)

antiochp committed:
cleanup unused data_file_path (#2766)

garyyu committed:
use a grin specific tmp dir instead of a system tmp dir (#2753) (#2772)

* use a grin specific tmp dir instead of a system tmp dir

* rustfmt

* using create_new instead of create

* switch the txhashset sandbox from system tmp dir to grin specific tmp dir

* rustfmt

* adapter get_tmp_dir delegate to chain implementation

* simply the logic of protolc for get_tmpfile_pathname

* rustfmt

* fix the merge mistake

DavidBurkett committed:
Increasing thread sleeps for socket listener thread and per-connection threads. (#2773)

hashmap committed:
Fix unused import warnings (#2775)

antiochp committed:
Relax our p2p rules around matching protocol versions (#2764)

antiochp committed:
Bump the version of croaring to 0.3.9 (#2763)

antiochp committed:
Cleanup how we deal with fixed size vs variable size data files (#2757)

introduce a SizeInfo enum (to replace two options)

ignopeverell committed:
Increase peer max count significantly (#2754)

* Increase peer max count significantly

* Reduce limit to 125, like bitcoin

hashmap committed:
Fix peer dropping (#2780)

It turns out that we drop connection if we fail to process a message
because of chain/store/internal error, eg we have a header already, so
we refuse it and drop the peer.
This pr doesn't forward this error to the peer error channel so the
connection will not be dropped.

hashmap committed:
Set TUI FPS (#2783)

New version of Cursive allows to set fps, the existing  set_autorefresh sets 30 fps. This change decreases CPU usage from 50-65% to 15-25% on my machine.

antiochp committed:
bucket_transactions is now fee_to_weight aware (#2758)

* bucket_transactions is now fee_to_weight aware
bucket_transactions now returns the underlying txs

* cleanup

* simplify pool bucket sorting, no need for depth sort key

antiochp committed:
Bump rand crate to 0.6 (was 0.5) (#2790)

shuffle was deprecated, fix this up

antiochp committed:
Add debug logging on startup to provide size of our leaf_set (aka UTXO set) (#2788)

ignopeverell committed:
i2p client connection impl; fix existing p2p tests

ignopeverell committed:
rustfmt

chisa0a committed:
P2P: fixups for 1.1.0

chisa0a committed:
Servers: fixups for 1.1.0

chisa0a committed:
P2P: remove panic!, add PeerAddrType enum

Convert use of panic! to return result + error.

Adds PeerAddrType enum for converting address type to/from raw u8.

chisa0a committed:
P2P: rustfmt + add I2P as fullnode capability

Adds I2P to fullnode capabilities, modifies tests accordingly.

Other rustfmt changes.

chisa0a commented:
Will close this for now, until I can get two nodes talking over I2P.

chisa0a commented:
Removed the rebase onto `1.1.0`, and adds remaining work for basic I2P connections. Some work remains, but this branch is now capable of making a basic connection over I2P.

@ignopeverell review ping.
